### ISSUE TITLE
Android lint errors when ca/fr/lt/sk are missing "many" in plurals, Weblate does not

### ISSUE THREAD

**Issue by eighthave** [WeblateOrg/weblate#7520]
### Describe the issue

Android's _lint_ tool is complaining about missing "many" entries in plural strings for "cs" (Czech), "fr" (French) , "lt" (Lithuanian), and "sk" (Slovak).  The Weblate interface does not offer "many" as an option for those languages/strings.  For example:

https://gitlab.com/grote/fdroidclient/-/jobs/2351872888
```
    ^
/builds/grote/fdroidclient/app/src/main/res/values-cs/strings.xml:396: Error: For locale "cs" (Czech) the following quantities should also be defined: many [MissingQuantity]
    <plurals name="details_last_update_months">
    ^
/builds/grote/fdroidclient/app/src/main/res/values-cs/strings.xml:401: Error: For locale "cs" (Czech) the following quantities should also be defined: many [MissingQuantity]
    <plurals name="details_last_update_years">
    ^
/builds/grote/fdroidclient/app/src/main/res/values-lt/strings.xml:406: Error: For locale "lt" (Lithuanian) the following quantities should also be defined: many [MissingQuantity]
    <plurals name="button_view_all_apps_in_category">
    ^
/builds/grote/fdroidclient/app/src/main/res/values-lt/strings.xml:501: Error: For locale "lt" (Lithuanian) the following quantities should also be defined: many [MissingQuantity]
    <plurals name="tts_view_all_in_category">
    ^
   Explanation for issues of type "MissingQuantity":
   Different languages have different rules for grammatical agreement with
   quantity. In English, for example, the quantity 1 is a special case. We
   write "1 book", but for any other quantity we'd write "n books". This
   distinction between singular and plural is very common, but other languages
   make finer distinctions.
   This lint check looks at each translation of a <plural> and makes sure that
   all the quantity strings considered by the given language are provided by
   this translation.
   For example, an English translation must provide a string for
   quantity="one". Similarly, a Czech translation must provide a string for
   quantity="few".
   https://developer.android.com/guide/topics/resources/string-resource.html#Plurals
```

This is related to #1550.  This site seems to say it is a matter of conversion between the CLDR and gettext representation schemes for plurals:
https://github.com/miracle2k/android2po#plurals-support

### I already tried

- [X] I've read and searched [the documentation](https://docs.weblate.org/).
- [X] I've searched for similar issues in this repository.

### Steps to reproduce the behavior

These are some of the strings:
* https://hosted.weblate.org/translate/f-droid/f-droid/lt/?offset=1&q=button_view_all_apps_in_category&sort_by=-priority%2Cposition&checksum=
* https://hosted.weblate.org/translate/f-droid/f-droid/cs/?offset=1&q=details_last_update_months&sort_by=-priority%2Cposition&checksum=

### Expected behavior

Weblate should support the "many" field for Android strings.xml for cs/fr/lt/sk languages.

### How do you run Weblate?

weblate.org service

**Comment by nijel** [comment]
This definitely was not the case in the past, what has changed in Android regarding this? Weblate uses CLDR data for Android.

 The `many` plural is used for decimal numbers (see https://unicode-org.github.io/cldr-staging/charts/37/supplemental/language_plural_rules.html#cs) and that indeed cannot be represented in Weblate right now.

**Comment by eighthave** [comment]
I'm not sure exactly when it changed, it could have been a while ago since I haven't been too active working on Android things recently.

Here's the lint check code:
https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/PluralsDetector.kt?q=%22the%20following%20quantities%20should%20also%20be%20defined%22

**Comment by nijel** [comment]
It was introduced in https://cs.android.com/android-studio/platform/tools/base/+/97f96e70464579c01712dec78bb398d2e47582df, unfortunately the commit does not include any reasoning...

**Comment by nijel** [comment]
Weblate certainly could be changed to deal with non-integer plurals. On the other side, most of the applications don't really need this, and it would add one additional string to translate (and confuse translators because they are currently not used to this).

**Comment by eighthave** [comment]
I have no opinion of whether it is needed or not.  I'm just reporting that 
Android is calling it a requirement, and they added a _lint_ check to enforce it.

**Comment by eighthave** [comment]
So here's more info on the use case:

* Dev who does not know anything about ca/fr/lt/sk puts their Android app
   on Weblate.
* Dev sets up standard Android _lint_ checks since the include important checks
   to prevent crashes due to incomplete translations.
* Dev has no idea what "many" plurals are, and wonders why _lint_ is failing.
* Dev has no idea how to fix it, so disables _lint_ or uses different
   translation service.

**Comment by warting** [comment]
maybe make it configurable in weblate what kind of different plurals it exists for each language?

**Comment by nijel** [comment]
The plural rules are currently detected based on the file. Some formats do explicitly include it (Gettext), for others we do guess what should be used (what turns out to be tricky, see for example https://github.com/translate/translate/issues/786).

Having that user configurable would be possible, but I'm uncertain if it will address the issue - the choice here is between Android lint complains and translators having extra work which is not needed in most cases. Anyway, it's safe to assume that most people will not change the configuration :-).

**Comment by eighthave** [comment]
As far as I know, the lint rules are not configurable beyond enable/disable 
across the project.  I don't think that Weblate should be configurable for the 
Android strings.xml unless Android itself is.

Do you think this could be a bug in Android lint?  If so, I'm happy to report 
it, but I don't know the details of CLDR and things like that.

**Comment by nijel** [comment]
Apparently, it was [intentional change](https://cs.android.com/android-studio/platform/tools/base/+/97f96e70464579c01712dec78bb398d2e47582df) there. Knowing the reasons might help us understand that as well...

**Comment by eighthave** [comment]
Based on the info here, my guess is that they added support for `many` to support strings with decimal numbers in them.  Seems like just an incremental improvement to their localization support.  Their test example is in Czech, does it make sense?

```
                                        + "<plurals name=\"product_type_ch_adult_general_abonnement\">\n"
                                        + "  <item quantity=\"few\">%d dospělí (karta GA)</item>\n"
                                        + "  <item quantity=\"many\">%d dospělých (karta GA)</item>\n"
                                        + "  <item quantity=\"one\">%d Dospělý (GA travelcard)</item>\n"
                                        + "  <item quantity=\"other\">%d Dospělí/dospělých (GA travelcard)</item>\n"
                                        + "</plurals>"))
```

**Comment by nijel** [comment]
Yes, it makes grammatical sense when you need decimals. The problem is that you don't need them in numerous instances. In this case, you won't have two and half adults, as you usually don't cut them ;-).

Anyway, to support this in Weblate:

* the plural rules have to be expanded to handle decimals
* translate-toolkit needs to understand these as well
* translate-toolkit should be able to migrate from the old to new plurals (it currently stores them in the list, so `other` would become `many` and `other` would be untranslated
* additional plural rules need to be added for Android in Weblate
* migrate existing translations in the Weblate database (again, these are stored sequentially, so without migration the plurals would be offset)

**Comment by kzaikin** [comment]
@samofcorinth could you explain the reasoning behind change from `other` to `many` (the discussion above)?
For example, Android pluralization engine and translation services (weblate.org or crowdin.com) use `one|other` for French, while android linter requires it to have 'one|many' - that requires us to treat lint issues as false-positives.

**Comment by samofcorinth** [comment]
Android Lint gets its plural data from this page: https://unicode-org.github.io/cldr-staging/charts/latest/supplemental/language_plural_rules.html#fr
According to this page, French (and the other languages) indeed has the `many` plural form. The version of the table that added `many` was v38 (introduced in 2020), so most likely Weblate's data came from a version before this one.

**Comment by nijel** [comment]
French seems a bit different here - there is an additional `n % 1000000` condition for `many` which was indeed introduced in CLDR v38, see https://github.com/unicode-org/cldr/commit/4116d5414204244adfaf44e830b32271aae91e4d and https://unicode-org.atlassian.net/browse/CLDR-12010.

For cs/lt/sk the issue is described above - it affects decimal numbers only.

**Comment by nijel** [comment]
I've split out https://github.com/WeblateOrg/weblate/issues/7888 to handle CLDR 38+ plural changes.

**Comment by ashutoshgngwr** [comment]
@eighthave 
cc @nijel 

> So here's more info on the use case:
> * Dev who does not know anything about ca/fr/lt/sk puts their Android app on Weblate.
> * Dev sets up standard Android _lint_ checks since the include important checks to prevent crashes due to incomplete translations.
> * Dev has no idea what "many" plurals are, and wonders why _lint_ is failing.

* Dev has no idea how to fix it, so ~~disables _lint_ or uses different translation service~~ manually fixes the lint errors by replacing `other` keys with `many`.
* Dev gets a report from a user that the app is force closing on their device.
* Dev opens up the Crashlytics dashboard to find a relevant issue with the following stacktrace: `android.content.res.Resources$NotFoundException: Plural resource ID #0x7f100004 quantity=14 item=other`.
* Dev confirms with the user if their device is using French locale.
* Dev can prevent this fatal crash by replacing `many` keys with `other` for plurals in the French locale (Android Lint still reports this as an error).
* Dev gets mad at someone. Not sure who, but someone.

---
Please excuse my misplaced humour! I've just run into this issue, and it appears to me that Android Lint is at fault.

I started with the French locale. If I use the following, the app crashes fatally, however, Android Lint doesn't report it as an error.

```xml
<plurals name="test">
    <item quantity="one">One</item>
    <item quantity="few">Few</item>
    <item quantity="many">Many</item>
</plurals>
```

If I use the following, the app works as expected, but Android Lint reports it as an error.

```xml
<plurals name="test">
    <item quantity="one">One</item>
    <item quantity="other">Other</item>
</plurals>
```

I then proceeded to reproduce this behaviour for Czech (cs), Lithuanian (lt) and Slovak (sk) and was successful in doing so for all these locales. I will try to find a related issue on Google's Issue Tracker or open a new one asking for clarification later today.

**Comment by darioseidl** [comment]
Hi, we're a little confused, specifically about the French plurals.

We were using Weblate 4.14 with i18next JSON v4, and had singular/plural suffixes `_one` and `_many`. Now, we upgraded to the latest Weblate 5.8.4, and I thought that [this would introduce a new plural rule for French](https://github.com/WeblateOrg/weblate/issues/7888), such that there are `_one`, `_many`, and `_other`. However, I still see the rule `n > 1` for French, and I'm not sure if that refers to `_many` or `_other` now.

 @nijel could you be so kind and explain how (French) plurals are currently handled in Weblate, based on which CLDR version, using integer or decimal rules?

We also have scripts to synchronize and sort our JSON translation files, and we need to know the plural suffices for those scrips, so it would be great if we knew where Weblate is taking them from. For the script, we're extracting the suffixes from the cardinal integer rules from https://github.com/unicode-org/cldr-json/blob/main/cldr-json/cldr-core/supplemental/plurals.json, but that doesn't seem to match Weblate, because for French that results in `_one`, `_many`, and `_other`.

**Comment by nijel** [comment]
@darioseidl See https://github.com/WeblateOrg/weblate/discussions/13279

**Comment by nikclayton** [comment]
@nijel Just discovered this after a lint issue on a Weblate PR that updates the Slovak translation.

The PR is https://github.com/pachli/pachli-android/pull/1250, which includes this addition:

```
    <plurals name="notification_severed_relationships_summary_followers_fmt">
        <item quantity="one">%1$s odstránený sledujúci</item>
        <item quantity="few">%1$s odstránení sledujúci</item>
        <item quantity="other">%1$s odstránených sledujúcich</item>
    </plurals>
```

Note the `other` as one of the quantities. This triggers this lint check:

>  For locale "sk" (Slovak) the following quantity should also be defined: many (e.g. "10.0 dňa") 

However, if I look at the Weblate UI for this translation, https://hosted.weblate.org/translate/pachli/app-main/sk/?q=followers_fmt&sort_by=-priority%2Cposition&checksum=, the `other` translation is shown, but the UI text for it is "Slovak, Many"

Here's a screenshot:

![Image](https://github.com/user-attachments/assets/da0ec227-5579-4abe-b2c2-febea1f2b538)

So what's shown in the UI as "Slovak, Many" is converted to `<item quantity="other">` in the output.

Not sure whether the bug here is:

- the content of the "Slovak, Many" field should be written as `<item quantity="many">`
- "Slovak, Many" is misnamed in the UI, and should be "Slovak, Other", and a new "Slovak, Many" UI field should be introduced

or something else entirely.

**Comment by emmapeel2** [comment]
We are having this issue with Slovak and Czech, see https://gitlab.torproject.org/tpo/applications/vpn/-/merge_requests/234#note_3211366 

The Android linter does not accept Weblate's translations, but we have no errors in Weblate, and the missing plurals are not visible on Weblate's interface.

**Comment by kewisch** [comment]
It appears the real fix for this issue is a bit more involved. Could there be a hack in the meanwhile to just use the same string for "many" and "other" in the affected locales and output both? While it might not be 100% correct, it would at least not crash the application.

**Comment by nijel** [comment]
It should be possible to implement such a workaround inside translate-toolkit. ~~Does anybody has a full list of affected locales?~~

PS: https://github.com/WeblateOrg/language-data/pull/1920 extracts it from the CLDR data.

**Comment by nijel** [comment]
https://github.com/translate/translate/pull/5641 implements this workaround.

Still, this is not a proper solution. In the long term, the translate-toolkit should expose plurals in a more structured way (or maybe switch to `moz-l10n` because they already do this) so that the logic does not have to be in two places, making any changes to it problematic.

**Comment by kewisch** [comment]
Thank you for implementing the workaround, so much appreciated!

**Comment by KwadroNaut** [comment]
~~Seems to work, thanks. I re-added Czech and Slovak to test.~~
I take that back, Slovak failed.
```
  34 actionable tasks: 34 executed
  /builds/kwadronaut/vpn/app/src/main/res/values-sk/strings.xml:90: Error: For locale "sk" (Slovak) the following quantity should also be defined: many (e.g. "10.0 d?a") [MissingQuantity]
      <plurals name="n_bridges">
      ^
```
https://gitlab.torproject.org/kwadronaut/vpn/-/jobs/1059240#L138

fwiw, in the xml:

```
    <plurals name="n_bridges">
        <item quantity="one">%1$s most pridaný</item>
        <item quantity="few">%1$s mosty pridané</item>
        <item quantity="other">%1$s mostov pridaných</item>
    </plurals>
```

**Comment by TheLastProject** [comment]
I'm not sure if this change is live on your official Hosted Weblate, if not, ignore the below:

Tried to re-enable the missingQuantity lint as a test. I'm seeing it fail for Hebrew: https://github.com/CatimaLoyalty/Android/actions/runs/16508037966/job/46683618889?pr=2580

```
/home/runner/work/Android/Android/app/src/main/res/values-iw/strings.xml:4: Error: For locale "iw" (Hebrew) the following quantity should also be defined: many (e.g. "20 שנה") [MissingQuantity]
    <plurals name="selectedCardCount">
    ^
```

I can confirm that translation is fully filled in on Weblate: https://hosted.weblate.org/translate/catima/catima/he/?checksum=f2122da997dfe989&q=selectedCardCount&sort_by=-priority%2Cposition

**Comment by nijel** [comment]
There was no recent change in Hebrew (this specific change was for other languages; see https://github.com/translate/translate/pull/5641/files#diff-7b8b6b3d040ee64d2e90b7a9676e62a09d4ace818a3e54893282e8e9da8c4ea9R1011). 

The CLDR specification doesn't define the `many` category for Hebrew: http://www.unicode.org/cldr/charts/44/supplemental/language_plural_rules.html#he

If you figure out why Android wants `many` there, please open a separate issue; it appears to be different from the decimals handling described in this issue.

**Comment by kewisch** [comment]
It appears to me that for Hebrew there was a change between CLDR 41 and 42 that removed `many`. The linter appears to use either 36, 38, 41 or 42 based on target sdk version. 41 is for Tiramisu (13), 42 for anything above. Could someone test if updating the target SDK version helps?

**Comment by TheLastProject** [comment]
My case is with an app using targetSdk 35 (Android 15) so it definitely does happen on Android 15.

**Comment by nijel** [comment]
Please do not make this a catch-all issue for all CLDR/Android locale handling changes and keep this issue focused on decimal plurals only. These are currently worked around by duplicating content from `other` plural without an option to edit it in Weblate. Anything else should get a separate issue.

**Comment by grote** [comment]
Getting the same for Hebrew now:
```
src/main/res/values-he/strings.xml:662: Error: For locale "he" (Hebrew) the following quantity should also be defined: many [MissingQuantity]
      <plurals name="notification_summary_app_updates">
```

The weblate webinterface doesn't have the `many` option.

**Comment by eighthave** [comment]
I usually just manually copy one of the other strings, like `other` and call it 
`many.  You could check the Weblate issues, and file one if there isn't one already.

**Comment by nijel** [comment]
@grote This has been discussed here and in https://github.com/orgs/WeblateOrg/discussions/16015 and https://github.com/WeblateOrg/weblate/issues/7520#issuecomment-3244269100 still applies. Hiding all of the comments as off-topic.

TLDR: The expectance of `many` for Hebrew in the linter depends on the target SDK version because of changed definitions in CLDR, using 14+ will make it not required.

**Comment by kwunlokng** [comment]
You can catch CLDR plural rule violations at build time with `i18n-validate`:

```bash
npx @i18n-agent/i18n-validate values-cs/strings.xml
# ERROR: Czech requires: one, few, many, other
# Found: one, other — Missing: few, many

i18n-validate --dir res/ --strict --fail-on-error
```

Supports 32 formats, runs 10 validations including CLDR plural completeness. Outputs terminal, JSON, or JUnit XML for CI.

Install: `npm install -g @i18n-agent/i18n-validate`

https://github.com/i18n-agent/i18n-validate
